siobrultech_protocols/gem/protocol.py

```diff
@@ -15,16 +16,61 @@
 LOG = logging.getLogger(__name__)
 
 PACKET_HEADER = bytes.fromhex("feff")
+API_RESPONSE_WAIT_TIME_SECONDS = 3.0  # Time to wait for an API response
+PACKET_DELAY_CLEAR_TIME_SECONDS = 3.0  # Time to wait after a packet delay request so that GEM can finish sending any pending packets
 
 
 class PacketProtocol(asyncio.Protocol):
-    def __init__(self, queue: asyncio.Queue):
+    def __init__(
+        self,
+        queue: asyncio.Queue,
+        on_connection_made: Optional[Callable[[GemApi], None]] = None,
+    ):
         self._buffer = bytearray()
         self._queue = queue
-        self._transport: Optional[asyncio.BaseTransport] = None
-
-    def connection_made(self, transport: asyncio.BaseTransport):
+        self._transport: Optional[asyncio.WriteTransport] = None
+        self._api_lock = asyncio.Lock()
+        self._api_mode = False
+        self._on_connection_made = on_connection_made
+
+    async def _send_api_command(self, command: str):
+        async with self._api_lock:  # One API call at a time, please
+            if not self._transport:
+                raise EOFError
+
+            # We're about to send a request on the same channel that the GEM is using to
+            # push data packets to us. To minimize confusion, we ask the GEM to delay packets
+            # for 15 seconds and give it a few seconds to finish sending any in-progress
+            # packets before sending our request.
+            LOG.debug("Requesting packet delay...")
+            self._transport.write(
+                DELAY_NEXT_PACKET.encode()
+            )  # Delay packets for 15 seconds
+            await asyncio.sleep(PACKET_DELAY_CLEAR_TIME_SECONDS)
+
+            if not self._transport:
+                raise EOFError
```

Comment:
We have this in two places in this method, plus in `_get_packet`.  I think it might be better, at this point, to make a small helper (which will be more clear of purpose as well):

```python
  def _ensure_transport_alive(self) -> None:
    if not self._transport:
      raise EOFError
```

Reply:
In this method the checks serve to tell PyLance that the field is not `None`, and it's not smart enough to see that a call to `_ensure_transport_alive` does that. I'll factor out a similarly-named helper that returns the instance.